x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java

```diff
@@ -660,12 +684,42 @@ public PlanFactory visitJoinCommand(EsqlBaseParser.JoinCommandContext ctx) {
                 }
                 joinFields.add(ua);
             } else {
-                throw new ParsingException(f.source(), "JOIN ON clause only supports fields at the moment, found [{}]", f.sourceText());
+                throw new ParsingException(
+                    f.source(),
+                    "JOIN ON clause only supports fields or AND of Binary Expressions at the moment, found [{}]",
+                    f.sourceText()
+                );
+            }
+        }
+        validateJoinFields(joinFields);
+        return new JoinInfo(joinFields, emptyList());
+    }
+
+    @Override
+    public JoinInfo visitExpressionBasedLookupJoin(EsqlBaseParser.ExpressionBasedLookupJoinContext ctx) {
+        var predicates = visitList(this, ctx.comparisonExpression(), Expression.class);
+        List<Attribute> joinFields = new ArrayList<>(predicates.size());
+        List<Expression> joinExpressions = new ArrayList<>(predicates.size());
+        for (var f : predicates) {
+            f = handleNegationOfEquals(f);
+            if (f instanceof EsqlBinaryComparison comparison
+                && comparison.left() instanceof UnresolvedAttribute left
+                && comparison.right() instanceof UnresolvedAttribute) {
+                joinFields.add(left);
```

Comment:
Looks like we assume the left field to be special. Do we have to? During resolution we should be able to determine which side a field comes from.

It'd be a tad simpler if the field based join just used the fields, and the condition based join just used the condition - leaving a `null` in the respective other part of the `JoinInfo` (and later downstream, `JoinConfig`).